docs/tools.md

```diff
@@ -0,0 +1,328 @@
+        }
+    }
+    ```
+  * **ToolBox Execution:** The target `ToolBox` receives the message, runs the **`call_tool`** logic (which includes validation), and emits a response message.
+  * **ToolBox Response Message (Success):**
+    ```json
+    {
+        "type": "ovos.persona.tools.<toolbox_id>.call.response",
+        "data": {
+            "toolbox_id": "web_search_tools",
+            "result": {"field1": "output_value", "field2": 456} // Raw dictionary output
+        }
+    }
+    ```
+  * **ToolBox Response Message (Error):**
+    ```json
+    {
+        "type": "ovos.persona.tools.<toolbox_id>.call.response",
+        "data": {
+            "toolbox_id": "web_search_tools",
+            "error": "ValueError: Invalid input for 'tool_name'..."
+        }
+    }
+    ```
+
+-----
+
+## Implementation Guide
+
+This guide provides the steps for implementing a concrete **ToolBox** plugin and includes a toy example for file operations.
+
+To create a new set of discoverable tools, you must implement a class that inherits from `ToolBox` and fulfill its abstract requirements, primarily the `discover_tools` method.
+
+### Step 1: Define Input and Output Schemas
+
+For each tool, you must first define its data contract using Pydantic models. These models must inherit from `ToolArguments` or `ToolOutput` and define the expected fields.
+
+```python
+from pydantic import Field
+from typing import Optional
+from ovos_plugin_manager.templates.agent_tools import ToolArguments, ToolOutput
+
+# 1. Input Schema for Reading
+class FileReadArguments(ToolArguments):
+    """Arguments for the read_file tool."""
+    path: str = Field(description="The absolute path to the file to be read.")
+    encoding: str = Field(default="utf-8", description="The character encoding of the file (e.g., 'utf-8').")
+
+# 2. Output Schema for Reading
+class FileReadOutput(ToolOutput):
+    """Output structure for the read_file tool."""
+    content: str = Field(description="The full content of the file as a string.")
+    file_size_bytes: int = Field(description="The size of the file in bytes.")
+    success: bool = Field(default=True)
+
+# 3. Input Schema for Writing
+class FileWriteArguments(ToolArguments):
+    """Arguments for the write_file tool."""
+    path: str = Field(description="The absolute path to the file to be written or overwritten.")
+    content: str = Field(description="The string content to write to the file.")
+    append: bool = Field(default=False, description="If True, append to the file; otherwise, overwrite/create it.")
+
+# 4. Output Schema for Writing
+class FileWriteOutput(ToolOutput):
+    """Output structure for the write_file tool."""
+    path: str = Field(description="The path to the file that was written.")
+    bytes_written: int = Field(description="The number of bytes successfully written.")
+    success: bool = Field(default=True)
+    error: Optional[str] = None
```

Comment:
Related to my question above, seems that there are several ways to go about handling valid outputs vs. errors. Might be good to standardize on one for consistency